### ISSUE TITLE
[FR] Adjust Z probe offset based on temperatures

### ISSUE THREAD

**Issue by KaSroka** [MarlinFirmware/Marlin#17092]
### Description
I have noticed that my perfect Z probe offset varies with bed and nozzle temperature. It would be nice to allow for temperature compensation of this value.

Those are my measurements:

Bed temp | Nozzle temp | Perfect Z probe offset
-- | -- | --
50 | 200 | 1,075
70 | 240 | 1,025
110 | 240 | 1

From my measurements I've got those coefficients: 
Bed temp coefficient -0.00062
Nozzle temp coefficient 0.000938

### Feature Workflow
We may add new parameters to the configuration:
```
#define Z_PROBE_OFFSET_BED_TEMP_COEFF -0.00062
#define Z_PROBE_OFFSET_NOZZLE_TEMP_COEFF 0.000938
#define Z_PROBE_OFFSET_BED_TEMP_CALIB 110
#define Z_PROBE_OFFSET_NOZZLE_TEMP_CALIB 240
```

And adjust probe Z offset based on current temperatures:
`adjusted_z_offset = z_offset + (Z_PROBE_OFFSET_NOZZLE_TEMP_CALIB - current_nozzle_temp) * Z_PROBE_OFFSET_NOZZLE_TEMP_COEFF + (Z_PROBE_OFFSET_BED_TEMP_CALIB - current_bed_temp) * Z_PROBE_OFFSET_BED_TEMP_COEFF`

**Comment by swilkens** [comment]
The default for the coefficient would probably have to be "1" or the entire feature may have to be off by default, just to ensure users with beds that are less / non-reactive in geometry to heat that level the bed at room temp don't suddenly have a surprise deviation at printing temperature.

I like the FR though, it makes sense to account for bed warping due to temperature increase - especially on the thinner beds. It might even be worked into UBL @Roxy-3D ?

Though I wonder if these CTE geometry changes are within the error range of typical probes, in that case it's problematic.

**Comment by KaSroka** [comment]
Those measurements I've got are consistent (tested on at least 4 prints each). I'm using BL-touch from Ali. And I've verified that first layer is always 0.1mm (+/- 0.01mm)

The default should be 0 but I think that disabling this feature by default is a better option.

**Comment by davidcgu** [comment]
try a pinda, it's already implemented since 2.0,.3

**Comment by tombrazier** [comment]
The probe temperature compensation code (see PROBE_TEMP_COMPENSATION in Configuration_adv.h) could easily be modified to work with a case where there is only a bed and extruder temperature sensor. That would cover this FR.

**Comment by tombrazier** [comment]
I could code this if anyone wants. I think it's probably a simple matter of putting #if clauses around the different bits of probe temperature compensation code so that only the bits related to the actual temperature sensors you have get compiled. The config would then change from the following...

```
#if HAS_BED_PROBE && TEMP_SENSOR_PROBE && TEMP_SENSOR_BED
  // Enable thermal first layer compensation using bed and probe temperatures
  #define PROBE_TEMP_COMPENSATION
...
#endif
```

to

```
// Enable thermal first layer compensation using bed and probe temperatures
//#define PROBE_TEMP_COMPENSATION
#if ENABLED(PROBE_TEMP_COMPENSATION)
...
#endif
```
and HAS_BED_PROBE, TEMP_SENSOR_PROBE and TEMP_SENSOR_BED will be used to omit code in files like the following:

Marlin/src/gcode/calibrate/G76_M192_M871.cpp
Marlin/src/feature/probe_temp_comp.cpp
Marlin/src/feature/probe_temp_comp.h

**Comment by tombrazier** [comment]
I think this issue can now be closed.

#23033 is now merged into the bugfix-2.0.x branch. The following config (in Configuration_adv.h) will now give the behaviour requested in the OP:

```
#define PTC_BED      // Compensate based on bed temperature
#define PTC_HOTEND   // Compensate based on hotend temperature
```
and
```
  #if ENABLED(PTC_BED)
    // Bed temperature calibration builds a similar table.
    #define PTC_BED_START     110    // (°C)
    #define PTC_BED_RES         1    // (°C)
    #define PTC_BED_COUNT       1
    #define PTC_BED_ZOFFS     { -620 } // (µm) Z adjustments per sample
  #endif

  #if ENABLED(PTC_HOTEND)
    // Note: There is no automatic calibration for the hotend. Use M871.
    #define PTC_HOTEND_START 240    // (°C)
    #define PTC_HOTEND_RES     1    // (°C)
    #define PTC_HOTEND_COUNT   1
    #define PTC_HOTEND_ZOFFS  { 938 } // (µm) Z adjustments per sample
  #endif
```
